src/dfx/src/commands/canister/snapshot.rs

```diff
@@ -0,0 +1,187 @@
+use ic_utils::interfaces::management_canister::CanisterStatus;
+use indicatif::HumanBytes;
+use itertools::Itertools;
+use slog::info;
+use time::{macros::format_description, OffsetDateTime};
+
+use crate::lib::{
+    environment::Environment,
+    error::{DfxError, DfxResult},
+    operations::canister::{
+        delete_canister_snapshot, get_canister_status, list_canister_snapshots,
+        load_canister_snapshot, take_canister_snapshot,
+    },
+    root_key::fetch_root_key_if_needed,
+};
+
+#[derive(Parser)]
+pub struct SnapshotOpts {
+    #[command(subcommand)]
+    subcmd: SnapshotSubcommand,
+}
+
+/// Controls canister snapshots that can reset a canister to an earlier state of execution.
+#[derive(Subcommand)]
+enum SnapshotSubcommand {
+    /// Creates a new snapshot of a canister. The canister must be stopped.
+    Create {
+        /// The canister to snapshot.
+        canister: String,
+        /// If a snapshot ID is specified, this snapshot will replace it and reuse the ID.
+        #[arg(long)]
+        replace: Option<SnapshotId>,
+        /// Force snapshot creation even if the canister is running. Not recommended unless you know what you're doing.
+        #[arg(long, short)]
+        force: bool,
+    },
+    /// Loads a canister snapshot, overwriting its execution state. All data since that snapshot will be lost. The canister must be stopped.
+    Load {
+        /// The canister to load the snapshot in.
+        canister: String,
+        /// The ID of the snapshot to load.
+        snapshot: SnapshotId,
+        /// Force snapshot loading even if the canister is running. Not recommended unless you know what you're doing.
+        #[arg(long, short)]
+        force: bool,
+    },
+    /// Lists a canister's existing snapshots.
+    List {
+        /// The canister to list snapshots from.
+        canister: String,
+    },
+    /// Deletes a snapshot from a canister.
+    Delete {
+        /// The canister to delete the snapshot from.
+        canister: String,
+        /// The ID of the snapshot to delete.
+        snapshot: SnapshotId,
+    },
+}
+
+#[derive(Clone)]
+struct SnapshotId(Vec<u8>);
+
+impl Display for SnapshotId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str(&hex::encode(&self.0))
+    }
+}
+
+impl FromStr for SnapshotId {
+    type Err = DfxError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self(hex::decode(s)?))
+    }
+}
+
+pub async fn exec(
+    env: &dyn Environment,
+    opts: SnapshotOpts,
+    call_sender: &CallSender,
+) -> DfxResult {
+    fetch_root_key_if_needed(env).await?;
+    let canisters = env.get_canister_id_store()?;
+    let logger = env.get_logger();
+    match opts.subcmd {
+        SnapshotSubcommand::Create {
+            canister,
+            replace,
+            force,
+        } => {
+            let canister_id = canister.parse().or_else(|_| canisters.get(&canister))?;
+            if !force {
```

Comment:
Where is this "force" behavior coming from a design/spec/product perspective? There is a race condition here as long as dfx enforces the behavior.

Reply:
My assumption was that someone would *eventually* find a use for it. Is it niche enough that we should not support it? And I don't think there's any way to remove the race condition unfortunately with the current protocol.

Reply:
I don't think we should support a `force` flag. At least an enforced order of operations `dfx canister stop && dfx canister snapshot create/apply && dfx canister start` gives a hint that it might not be good if another `dfx canister start` happens too soon.

Unfortunately, in either case, it would be reasonable to assume that the replica performs this kind of integrity check, since it's the only way to actually enforce the constraint. I would also recommend sending this back to the feature designers to find out if there is a good reason why it isn't so.